### ISSUE TITLE
Variable Product Parent Price Sync Not Always Ran

### ISSUE THREAD

**Issue by ObliviousHarmony** [woocommerce/woocommerce#24947]
**This bug report requires the WooCommerce Product Vendors plugin to replicate, but I'm pretty sure it is a bug in core, not in the extension.**

**Describe the bug**
When you submit a product for review, the `_price` meta for the parent product is not set correctly, leading to an invalid state once the product is published using quick or bulk edit options. If you save it using the publish button on the edit page, the error does not occur. I haven't really investigated why, since it wasn't broken.

The consequences of this are that `WC_Product::is_purchasable()` returns false due to the empty price meta on the parent product, leading to some strange effects. One specific example is that the button in the storefront reads "Read More" instead of "Select Options".

**To Reproduce**
1. Log in as a Vendor Manager user.
2. Create a new variable product.
3. Add an attribute to be used with variations that have a single option.
4. Add a variation for the attribute and set the regular price.
5. Submit the variation for review. If you look at the product listing now, you should see not see any price in the price column, this is the source of the bug. **Sometimes I found if I went through the steps too quickly this wouldn't happen, I'm not really sure why. If it doesn't, just start over.**
6. Log in as an administrator.
7. Using quick edit (or bulk edit) and set the status of the product to "Published"
8. See the "out of stock" message, despite the variation not having any stock problems.
9. Check the storefront and you'll see "Read More" for the product even though it is in stock.

You can then fix the issue by going into the product as the admin, changing the price on the variation, and saving it. If you look, the message will be gone, and it will work as expected.

**Troubleshooting**
I think I've tracked the bug down to  two possible issues.

The first is caused by the nonce in `WC_Admin_Post_Types::bulk_and_quick_edit_save_post`. This nonce is not sent when submitting the post using the "Submit for Review", "Save Draft", "Publish", and "Update" buttons in the interface. A consequence of this is that the variable product sync is not executed and the `_price` meta is not set. In my testing, I commented this out, which led to the next one.

The second issue is in `WC_Product_Variable_Data_Store_CPT::sync_price`. I haven't really looked into WHY, but for some reason the product's `get_visible_children()` method doesn't always return all of the visible variations? In my testing I found that when paired with the above steps, the review submission and publishing steps have an empty array instead of the expected result. All other times the method is called it is fine. To verify that this was an issue, I changed it to `get_children()` for testing purposes, and found that the issue was resolved entirely.

**Screenshots**
https://cld.wthms.co/GwEfcj

**Expected behavior**
When the product is published it should update the `_price` meta according to the children, preventing it from being an empty string.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
`
### WordPress Environment ###

WordPress address (URL): http://woo.test
Site address (URL): http://woo.test
WC Version: 3.7.1
REST API Version: ✔ 1.0.2
Log Directory Writable: ✔
WP Version: 5.2.4
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.17.5
PHP Version: 7.2.24-1+ubuntu18.04.1+deb.sury.org+1
PHP Post Max Size: 1 GB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.58.0
OpenSSL/1.1.1d

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.3.18-MariaDB-1:10.3.18+maria~bionic-log
Max Upload Size: 1 GB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.7.1
WC Database Prefix: wp_
MaxMind GeoIP Database: ✔
Total Database Size: 2.72MB
Database Data Size: 1.72MB
Database Index Size: 1.00MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 1.03MB + Index: 0.05MB + Engine InnoDB
wp_postmeta: Data: 0.06MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.05MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wcpv_commissions: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wcpv_per_product_shipping_rules: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (4) ###

JSON Basic Authentication: by WordPress API Team – 0.1
Jetpack by WordPress.com: by Automattic – 7.8
WooCommerce Product Vendors: by WooCommerce – 2.1.21
WooCommerce: by Automattic – 3.7.1

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #6 - /?page_id=6
Cart: #7 - /?page_id=7
Checkout: #8 - /?page_id=8
My account: #9 - /?page_id=9
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 2.5.3
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Complete: 1
Oldest: 2019-10-30 19:26:47 +0000
Newest: 2019-10-30 19:26:47 +0000

Pending: 0
Oldest: –
Newest: –

Canceled: 0
Oldest: –
Newest: –

In-progress: 0
Oldest: –
Newest: –

Failed: 0
Oldest: –
Newest: –

`
</details>

**Comment by juliaamosova** [comment]
@ObliviousHarmony thanks for the detailed report. However, I am not able to reproduce the behavior you are seeing. I have followed the steps you listed to replicate the issue. I've tried both to go slow and fast through the steps. 

Below is a screencast of what I am getting in either case (Admin on the left side of the screen, vendor on the right) - it works as expected: 

https://recordit.co/cFS3NmfzLq

Let me know if I missed any steps. 

cc @woocommerce/proton to see if someone else can reproduce the behavior described.

**Comment by ObliviousHarmony** [comment]
@juliaamosova Ah, so you're supposed to use the quick edit or bulk edit options. When you go directly to the product page and click "publish" it doesn't happen for some reason.

**Comment by juliaamosova** [comment]
@ObliviousHarmony thanks for additional details. I can now reproduce the issue. See screencast I made:

https://recordit.co/EO29RRvdS5

In addition to the video, the product shows as `Out of stock` on the vendor's dashboard as well but when opening the product's page, it shows that it is in stock. 

Admin:

![admin1](https://user-images.githubusercontent.com/19143190/68387439-3a96d480-0156-11ea-88f5-0a8e3fcd9846.jpg)

![admin2](https://user-images.githubusercontent.com/19143190/68387445-3e2a5b80-0156-11ea-9b92-b7fe50dd8235.jpg)

Vendor:

![vendor1](https://user-images.githubusercontent.com/19143190/68387451-42ef0f80-0156-11ea-89d8-b8908ae289f0.jpg)

![vendor2](https://user-images.githubusercontent.com/19143190/68387462-471b2d00-0156-11ea-8155-4a7f7edad7ed.jpg)

**Comment by rodelgc** [comment]
As this bug is already quite dated, we'll try to reproduce this on the current released version of WooCommerce to see if this issue is still happening. I'm adding the `status: needs reproduction` label for the meantime.

**Comment by swatipawarGS** [comment]
Hi @ObliviousHarmony ,

hank you for taking the time to report this bug, we really appreciate your help. We are unable to reproduce the issue on our end using WooCommerce 6.0.0 version and WordPress 5.8.2 version.


Please find below screencast  for reference :

https://user-images.githubusercontent.com/94531721/146381549-bac12ad5-fd87-43d9-95cd-961d05fc5cc8.mp4

Please provide us more detailed and any additional steps required to repro this issue which may help us to evaluate it further.

**Comment by rodelgc** [comment]
Thanks @swatipawarGS,

It also seems that this issue is no longer happening when I tested it myself. As seen from the screen video below, the product appears correctly as "In stock" and I was able to add it to my cart successfully, instead of getting the "Read more" notification.

https://user-images.githubusercontent.com/4509348/146733945-0505fcbf-b53a-4bdc-8ed8-ec7fcdeebae1.mov


This can probably be closed off?
